### ISSUE TITLE
UnrealIRCd NAMES vs WHO mode symbols cause AssertionError

### ISSUE THREAD

**Issue by half-duplex** [sopel-irc/sopel#1482]
```
>>1550121698.9370964    JOIN #test
<<1550121705.275868     :sopel!~sopel@bothost JOIN :#test
<<1550121705.2780008    :alpha.example.com 353 sopel = #test :sopel ~@me
<<1550121705.2786467    :alpha.example.com 366 sopel #test :End of /NAMES list.
<<1550121714.5684164    :alpha.example.com 352 sopel #test ~me myhost alpha.example.com me Hr~ :0 me
<<1550121714.5691297    :alpha.example.com 315 sopel #test :End of /WHO list.
<<1550121730.9563427    :me!~me@myhost PRIVMSG sopel :.exec bot.privileges["#test"]["me"]
>>1550121730.958703     PRIVMSG me :20
<<1550121741.7689545    :me!~me@myhost PRIVMSG sopel :.exec bot.channels["#test"].privileges["me"]
>>1550121741.7704542    PRIVMSG me :16
```

`Sopel 6.6.2 (running on python 3.7.2)`
`Unreal3.2.10.6`

It seems UnrealIRCd's RPL_NAMREPLY (353) reports all channel mode symbols (`~@me`, OP and OWNER) but its RPL_WHOREPLY (352) only gives the highest (`Hr~`, OWNER).
[handle_names](https://github.com/sopel-irc/sopel/blob/90a091614f7428fc6f419f5a50810850446b36f6/sopel/coretasks.py#L170) then sets `bot.privileges[][]` to 20 (OP|OWNER) and [_record_who](https://github.com/sopel-irc/sopel/blob/90a091614f7428fc6f419f5a50810850446b36f6/sopel/coretasks.py#L707) sets `bot.channels[].privileges[]` to 16 (OWNER).

It will explode the next time that user's modes are changed, with `AssertionError (file ".../sopel/coretasks.py", line 246, in track_modes)` since the two don't match.

**Comment by dgw** [comment]
Exactly the kind of thing that assertion was meant to catch. I'm glad of leaving it in the released code, now.

On the one hand, this reveals that `handle_names()` only updates `bot.privileges` and `_record_who()` only updates `bot.channels`.

On the other hand, fixing that inconsistency wouldn't fix _this_ bug. Sopel would still have incorrect mode masks stored, depending on which reply type it received last (gotta love race conditions).

@half-duplex's suggestion on IRC was to replace `bot.privileges` with a shim that translates accesses to it into `bot.channels` behind the scenes. That's a good idea, and probably should have happened a long time ago (i.e. when `bot.privileges` was first deprecated, in Sopel 6.2.0), but it's much more involved than would be reasonable to implement in time for the next bugfix release. That said, if someone wants to implement that, it's a good candidate for a future 6.6.x version (just not 6.6.3), or for Sopel 7.

For ASAP release, @half-duplex, would you make a PR adding in the missing set operations in the names/who functions you pointed out? Both should update both privileges data structures.

We'll have to have a hard think about how to handle NAMES vs. WHO mode symbols, though. I'm going to look at some RFCs, dig around in some ircd code (for UnrealIRCd and others), and try to make a decision from that on whether this is Sopel's bug or Unreal doing something wrong.

**Comment by dgw** [comment]
For the record, in [RFC 2812](https://tools.ietf.org/html/rfc2812), both `RPL_WHOREPLY` and `RPL_NAMREPLY` have the mode-character section defined as `"@" / "+"`, i.e. they both are meant to expect only one mode character (presumably the highest, or "most important", mode set).

Let's not worry about which is "correct" for now. Let's worry about consistency only, which is the point of having that `assert` in the first place. A future version of Sopel (7.x or beyond) can worry about making damn sure it gets all user modes on connect. Present-day Sopel (6.6.x) just wants to avoid exploding.

`track_modes()` uses `priv = priv & ~value` to remove modes, and `priv = priv | value` to add them. Both are safe to apply even if the mode being added is already present or the mode being removed is not present. The only worry is for users with multiple modes set (commonly channel owners with `+qo`, or superops with `+ao`), their privileges will become `0` if the higher mode is removed.

**tl;dr:** Fix both `_record_who()` and `handle_names()` to update both privileges data structures for 6.6.x, and then look at how Sopel can _definitely_ get all the modes on connect for a later release. (Also, probably, make `bot.privileges` a shim that accesses `bot.channels` instead of storing duplicate info.)

**Comment by half-duplex** [comment]
Fixed by #1494

**Comment by dgw** [comment]
Original reporter says this is fixed by #1494, so I'll close the issue.

@half-duplex Reopen this if you run into it again, OK? :smile_cat:
